File: AL_BogoSort.py
import random
import time
'''
Bogo Sort
Time complexity: O(N * (N!))

Inputing visualize = True as a second argument prints out array at
every outer loop iteration.

'''
def isSorted(arr):
	N = len(arr)

	for i in range(N-1):
		if arr[i] > arr[i+1]:
			return False

	return True
def BogoSort(arr, visualize=False):
	counter = 0

	while True:	
		counter += 1
		if (visualize):
				print(arr)

		random.shuffle(arr)

		if isSorted(arr):
			break
	if (visualize):
		print(arr)
	print("Number of shuffles: " + str(counter))



def testRun():
	random.seed(time.time())
	testN = 5

	testArr = [0]*testN

	for i in range(testN):
		testArr[i] = random.randrange(100)

	BogoSort(testArr, True)



testRun()

File: AL_BogoSort.cpp
#include <algorithm>
#include <iostream>
#include <random>
#include <string>
#include <vector>

#include <stdlib.h>
#include <time.h>
/* BubbleSort
 * Time complexity: O(N * (N!))
 */
template<typename T>
bool isSorted(std::vector<T> &vec) {
	for (uint i = 1; i < vec.size(); i++) {
		if (vec[i] < vec[i-1]) {
			return false;
		}
	}
	return true;
}

template<typename T>
void bogoSort(std::vector<T> &vec) {
	uint counter = 0;
	while (isSorted(vec) == false) {
		std::random_shuffle(vec.begin(), vec.end());
		counter++;
	}
	std::cout<<"Sorted after " << counter << " shuffles!" << std::endl;
	return;

}

template<typename T>
void print(std::vector<T> vec) {
	std::cout<<"[";
	for (uint i = 0 ; i < vec.size(); i++) {
		std::cout<<vec[i]<<" ";
	}
	std::cout<<"]"<<std::endl;
	return;
}

int myrandom (int i) { return std::rand()%i;}

void testRun() {
	int testN = 8;
	std::vector<int> testArr(testN);
	srand(time(0));

	for (uint i = 0 ; i < testArr.size(); i++) {
		testArr[i] = i;
	}

	std::random_shuffle(testArr.begin(), testArr.end(), myrandom);
	print(testArr);
	bogoSort(testArr);
	print(testArr);
	
}

int main() {

	testRun();
	return 0;
}